### PR TITLE
test(per-partition-limit): added new test

### PR DESCRIPTION
**Pull request by soyacz**
It is now possible to limit read rates and writes rates into a partition with a new WITH per_partition_rate_limit clause for the CREATE TABLE and ALTER TABLE statements. This is useful to prevent hot-partition problems when high rate reads or writes are bogus (for example, arriving from spam bots)
Feature introduced in:
https://github.com/scylladb/scylladb/commit/dab56b82fae5e36f7aa2ca6700d8cfa5baa2b515

Each cql request goes through per-partition limiter creating a risk of latency degradation.

Test scenario is about measuring basic performance metrics (latency, op/s) with and without per-partition-limit set and verifying the impact. Also measure influence of hot-partition scenario on these metrics.

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by soyacz** [comment]
Please review size of the cluster and load - I'm not sure if I set it correctly (maybe we should measure with higher load?)
Staging test job: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/per-partition-limit-test/3/

**Comment by soyacz** [comment]
Some cluster utilisation graphs for test run:
image
we can see 10m blocks on it:
1. base write
2. base read
3. base write + 20 partitions heavy load
4. base read + 20 partitions heavy load
5. limited write
6. limited read
7. limited write + 20 partitions heavy load
8. limited read + 20 partitions heavy load

Results (in logs) are showing no latency degradation and improvement of latency and op/s when additional heavy load for limited partition set is ongoing.

**Comment by soyacz** [comment]
Need to fix email report as it failing sending results.

**Comment by fruch** [comment]
> Some cluster utilisation graphs for test run: image we can see 10m blocks on it:
> 
> 1. base write
> 2. base read
> 3. base write + 20 partitions heavy load
> 4. base read + 20 partitions heavy load
> 5. limited write
> 6. limited read
> 7. limited write + 20 partitions heavy load
> 8. limited read + 20 partitions heavy load
> 
> Results (in logs) are showing no latency degradation and improvement of latency and op/s when additional heavy load for limited partition set is ongoing.

I'm not sure this causing a hot partition situation, change the monitor view to be per instance, and then par shard.
we should expect to see in the case of hot partition, one node, or one shard that gets most of the traffic.

i.e. the 20 partitions traffic should be the majority of the traffic. I would start doing just those on their own and analyze/measure it first.

looking at the issue that initiate this: https://github.com/scylladb/scylladb/issues/4703
you should be looking at the P99 latency, and not the throughput 

also I would separate it to multiple keyspaces, what you call base load should be on a different keyspace, and not limited


@piodul can you share the scylla-bench command you were using for testing ?

**Comment by piodul** [comment]
> @piodul can you share the scylla-bench command you were using for testing ?

Unfortunately I can't find them right now (I'm not sure I still have them, but I'll try to look later), but the things I for sure did differently were:

- I used larger concurrency over the scylla-bench invocation that was meant to cause overload (something like 1k or 2k).
- I overloaded only one partition (so that only one shard gets higher load).

Also, make sure that scylla-bench you use is shard-aware because rate limiting works best in that scenario.

**Comment by soyacz** [comment]
latency graphs:
image
we can see that latency during hot-partitions load without feature enabled is much higher than when we limit rate and is quite steady.

File: test-cases/features/per-partition-limit.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+test_duration: 180
+# using scylla-bench instead cassandra-stress as c-s rapidly retries on failure (e.g. due per-partition limit error)
+# causing overloading cluster due flood of failing requests
+
+stress_cmd_w: "scylla-bench -workload=sequential -mode=write -replication-factor=3 -partition-count=1000000 -clustering-row-count=1 -concurrency=50 -connection-count=10 -rows-per-request=1 -consistency-level=quorum -iterations=0 -timeout=300s -duration=10m"
+stress_cmd_r: "scylla-bench -workload=sequential -mode=read -replication-factor=3 -partition-count=1000000 -clustering-row-count=1 -concurrency=50 -connection-count=10 -rows-per-request=1 -consistency-level=quorum -iterations=0 -timeout=300s -duration=10m"
```
Comment (piodul) [inline test-cases/features/per-partition-limit.yaml:6]: AFAIK scylla-bench, by default, ignores all errors and doesn't limit the number of errors received. We probably want to stop these after receiving an error, right? In that case, maybe we should set `-error-at-row-limit 1` here, or the new `-error-limit 1`.

**Comment by piodul** [comment]
> image In the test normal load latency and throughput is compared and how hot partition load affects it. I take results from s-b as they are more accurate than on the graphs.

Did that test run had 1 or 20 hot partitions? (With 20, I would expect the load to be more balanced)

> 
> > i.e. the 20 partitions traffic should be the majority of the traffic. I would start doing just those on their own and analyze/measure it first.
> 
> concurrency and connection count for hot partitions traffic is higher than for normal load in this test - I assume this make it to be majority. We can see it on graphs I think.

Having larger concurrency than the normal load IMO makes sense, and that is what I did in my tests.

> > looking at the issue that initiate this: [scylladb/scylladb#4703](https://github.com/scylladb/scylladb/issues/4703) you should be looking at the P99 latency, and not the throughput
> 
> I added throughput verification to test also - this feature should also bring benefit in throughput under hot partition load. My mistake I didn't show graphs for it (fixed in comment above).

I agree that we should verify both latency and throughput. We have fixed concurrency here, so according to Little's law both throughput and latency (_average_ latency, at least, but ofc we are interested in p99 more) should become better.

**Review by piodul — DISMISSED** [review]
The test LGTM, only left some small nitpicks.

I initially misread the first message about the results and thought that per partition limiting didn't bring any improvement, so I suggested some changes. However, if the test shows improvements then I think the current parameters are perfectly fine.

I'm not sure whether having a single overloaded partition or multiple of them (so that all shards are loaded) is better, but I think both should be fine and should show that per-partition rate limiting increases the useful throughput of overloaded shards.

Increasing the load might not make sense, because you are using much weaker machines than I did.

To sum up, I think that the current parameters are fine.

File: test-cases/features/per-partition-limit.yaml (hunk 1 above)
Comment (fruch) [reply test-cases/features/per-partition-limit.yaml:6]: SCT is doing -error-at-row-limit 1000
If nothing else is defined.

But I agreed setting to fail on any error is reasonable here

**Comment by fruch** [comment]
> 
> > I'm not sure this causing a hot partition situation, change the monitor view to be per instance, and then par shard. we should expect to see in the case of hot partition, one node, or one shard that gets most of the traffic.
> 
> image
> In the test normal load latency and throughput is compared and how hot partition load affects it. I take results from s-b as they are more accurate than on the graphs.
> 
> > 
> > i.e. the 20 partitions traffic should be the majority of the traffic. I would start doing just those on their own and analyze/measure it first.
> 
> concurrency and connection count for hot partitions traffic is higher than for normal load in this test - I assume this make it to be majority. We can see it on graphs I think.
> 
> > 
> > looking at the issue that initiate this: [scylladb/scylladb#4703](https://github.com/scylladb/scylladb/issues/4703) you should be looking at the P99 latency, and not the throughput
> 
> I added throughput verification to test also - this feature should also bring benefit in throughput under hot partition load. My mistake I didn't show graphs for it (fixed in comment above).
> 
> > 
> > also I would separate it to multiple keyspaces, what you call base load should be on a different keyspace, and not limited
> > 
> 
> What will be benefit of that?
> 

It might help looking on each load seperatly,

But from the looks of those graphs is seems like it's clearer enough to see the results.

**Review by fruch — DISMISSED** [review]

**Comment by fruch** [comment]
@soyacz you might want to rebase to have the CI passing.

**Comment by soyacz** [comment]
rebased, please reapprove so we can merge.

**Review by piodul — APPROVED** [review]

**Review by fruch — APPROVED** [review]

**Comment by soyacz** [comment]
@roydahan I believe this should go to 2022.2 and 5.1 - do you approve?

**Comment by roydahan** [comment]
Yes, and to run it "officially" for 2022.2 at least once. (you can use 2022.2.0-rc6).

**Comment by soyacz** [comment]
Done. 
Test run: https://jenkins.scylladb.com/view/enterprise-2022.2/job/enterprise-2022.2/job/features-/job/features-per-partition-limit/4/
It failed in jenkins, but in reality it passed - error is due of flakiness of s-b in gathering results (showing +inf for op/s).
